33-green-wrasse/technical_specification.md

```diff
@@ -0,0 +1,340 @@
+- The access request is stored in the backend database
+- Notification email is sent out to DS
+- Confirmation email is sent out to DR
+
+When access was granted or denied:
+
+- DS logs in to Data Portal
+- DS views list of access requests
+- DS finds access request of DR
+- DS changes the status from "pending" to either "allowed" or "denied"
+- Notification email is sent out to DR
+- Confirmation email is sent out to DS
+- DR logs in to Data Portal
+- DR visits the download or profile page
+- DR should see the newly accessible dataset
+- DR can now create a work package for download.
+Creating work packages is covered by
+[epic 26](https://github.com/ghga-de/epic-docs/blob/main/26-dracula-ant/technical_specification.md).
+
+## API Definitions
+
+The Access Request Service REST API should have the following endpoints:
+
+Used by the web frontend to create and view access requests:
+
+- `POST /access-requests`
+  - auth header: internal access token
+  - request body:
+    - `user_id`: string (the registered user ID of the requester)
+    - `dataset_id`: string (the ID of the requested dataset)
+    - `work_type`: enum (download or upload, must be download for now)
+    - `email`: string (the contact email address of the requester)
+    - `request_text`: string (the text submitted with the request)
+    - `access_starts`: date (optional, when access should start)
+    - `access_ends`: date (optional, when access should end)
+
+The response body should include the ID of the newly created access request,
+although this is currently not used by the frontend.
+
+Used by the web frontend to fetch all existing requests:
+
+- `GET /access-requests`
+  - auth header: internal access token
+  - query parameters (optional):
+    - `dataset_id`: string (a dataset ID)
+    - `user_id`: string (a registered user ID)
+    - `state`: an access request state
+    - `work_type`: enum (download or upload)
+
+This endpoint gets the existing requests, filtered using the specified
+query parameters, and ordered by descending creation date.
+
+If the user specified in the access token does not have a data steward role,
+the `user_id` parameter is assumed to be the ID of that user,
+and an authorization error is returned if a different `user_id` is passed.
+
+The response body should be an array of Access Request Objects as described
+below.
+
+Note that currently we do not support parameters for filtering according to
+dates and limiting results or pagination, but these may be added later.
+
+Since the frontend first loads all relevant objects using this end point,
+these can be used in the detail view as well, and the backend does not need
+to provide a separate endpoint for getting individual Access Request Objects.
+
+Used by the web frontend to modify the status of an existing request:
+
+- `PATCH /access-requests/{access_request_id}`
+  - auth header: internal access token
+  - request body:
+    - `status`: enum (allowed/denied/pending)
```

Comment:
One could also model it as an optional boolean, where None means pending (no decision), `True` means approved, `False` means denied.

Reply:
Yes, but a boolean is more expressive, and we may want to add more status values later (like different stages of "pending" or "revoked" when it had been allowed and then later denied).